### PR TITLE
Better error message when there's no default IP pool

### LINKED ISSUE #4864: More detail in default ip-pool lookup failures
Recently when troubleshooting the cause of an instance failing to deploy, I saw that I was hitting an error that showed the `ip-pool` could not be found. I was looking the the logs to see which `ip-pool`, I only saw "Default pool for current silo" in the log message. Should we also provide the UUID or name being used in the query, so we actually know _what_ failed to be found?

https://github.com/oxidecomputer/omicron/blob/c794b1a33f6fbb500723b8c3ced9595b390f66a7/nexus/db-queries/src/db/datastore/ip_pool.rs#L187-L193

### PR DESCRIPTION
**Pull request by david-crespo**
Closes #4864 

This is a bad error message to get when the problem is that there is no default IP pool configured for your current silo:

```
not found: ip-pool with id "Default pool for current silo"
```

"Default pool for current silo" is not an id, so why would we call it one? This is better:

```
not found: default IP pool for current silo
```

This PR is just making that possible.

### CODE REVIEW

File: common/src/api/external/error.rs [hunk 1]
```diff
@@ -144,11 +144,11 @@ pub enum LookupType {
     ByName(String),
     /// a specific id was requested
     ById(Uuid),
-    /// a session token was requested
-    BySessionToken(String),
```
Comment (david-crespo) [inline common/src/api/external/error.rs:148]: This is not used. What's weird is it was introduced in #814, but it is also not used there.

File: common/src/api/external/error.rs [hunk 2]
```diff
@@ -144,11 +144,11 @@ pub enum LookupType {
...
     /// a specific id was requested with some composite type
     /// (caller summarizes it)
     ByCompositeId(String),
+    /// object selected by criteria that would be confusing to call an ID
+    ByOther(String),
```
Comment (david-crespo) [inline common/src/api/external/error.rs:151]: A catchall name like this is surely a code smell, but I can't think of a better name.

File: common/src/api/external/error.rs [hunk 3]
```diff
@@ -359,23 +359,22 @@ impl From<Error> for HttpError {
...
+                    LookupType::ByName(name) => {
+                        format!("{} with name \"{}\"", t, name)
                     }
+                    LookupType::ById(id) => {
+                        format!("{} with id \"{}\"", t, id)
+                    }
+                    LookupType::ByCompositeId(label) => {
+                        format!("{} with id \"{}\"", t, label)
+                    }
+                    LookupType::ByOther(msg) => msg,
```
Comment (david-crespo) [inline common/src/api/external/error.rs:372]: This is the point of all this — I need a variant that doesn't stick all that extra structure in the message.

File: nexus/db-queries/src/db/datastore/ip_pool.rs [hunk 4]
```diff
@@ -134,12 +134,8 @@ impl DataStore {
...
         //     .await?;
 
-        // join ip_pool to ip_pool_resource and filter
-
-        // used in both success and error outcomes
-        let lookup_type = LookupType::ByCompositeId(
-            "Default pool for current silo".to_string(),
-        );
+        let lookup_type =
+            LookupType::ByOther("default IP pool for current silo".to_string());
```
Comment (zephraph) [inline nexus/db-queries/src/db/datastore/ip_pool.rs:138]: Other than stripping the ID part of the string (which is helpful because it'll not longer look like an error) I don't know that this provides that much more actionable context. Something like this might be more helpful in the logs:

```rust
LookupType::ByQuery(format!("ip_pool where ip_pool_resource { resource_id: {}, is_default: true }", authz_silo_id));
```

It doesn't have to be exactly that. I know this doesn't communicate in plain english the error (as you've done above), but it does communicate some specifics of what tables to look at if something unexpected is happening. 

cc @internet-diglett do you have thoughts?

File: common/src/api/external/error.rs (hunk 1 above)
Comment (zephraph) [reply common/src/api/external/error.rs:148]: Thanks for removing this. I'm not sure how much sense it made in this context to begin with.

File: nexus/db-queries/src/db/datastore/ip_pool.rs (hunk 4 above)
Comment (david-crespo) [reply nexus/db-queries/src/db/datastore/ip_pool.rs:138]: I considered options more like that, giving the stuff we select by, but to a certain extent I think it’s revealing too much about how the concept of default pool is implemented. It feels weird for an API to be talking about a join table by name.

Comment (zephraph) [reply nexus/db-queries/src/db/datastore/ip_pool.rs:138]: How about combining the two? As with other errors in the system, there's both an internal and external message here. Internally we want enough context to know what to dig into if something's gone wrong. Externally we want a general message for the users to know what's failed at a high level.

Comment (david-crespo) [reply nexus/db-queries/src/db/datastore/ip_pool.rs:138]: True, interesting idea. Mostly only the 500-level errors have internal messages, except `InvalidValue` is a 400-level one used for things like disk size being too big and it does have that. It would be pretty invasive to change `ObjectNotFound` to have that.

https://github.com/oxidecomputer/omicron/blob/c392c76f0cc8a0bdaca5cf7690c6d95dd071e7db/common/src/api/external/error.rs#L28-L73

Comment (zephraph) [reply nexus/db-queries/src/db/datastore/ip_pool.rs:138]: Changing `ObjectNotFound` might be the right way to do it, but that's not what I specifically had in mind. 

image

In the logs above, there's the external event message as you'd expect but also a serialized version of the actual error object. Even if the "query" part of this never showed up in an error, we'd still see it in the logs (we, at least under the context at which this was initially seen). 

Now, it could be that we really only see errors serialized out like that in the logs under very special circumstances (maybe only saga events?). Were that the case, it may not be worth it.

**Comment by david-crespo** [comment]
Merging this as a net improvement. Will think about making it more better later.
